MSC2965: OAuth 2.0 Authorization Server Metadata discovery (#2965)
matrix-org/matrix-spec-proposals: proposals/2965-oidc-discovery.md

# MSC2965: OIDC Provider discovery

This proposal is part of the broader [MSC3861: Matrix architecture change to delegate authentication via OIDC](https://github.com/matrix-org/matrix-spec-proposals/pull/3861).

To be able to initiate an OAuth 2.0 login flow to use a Matrix server, the client needs to know the location of the authentication server in use.

## Proposal

Clients already discover the homeserver when doing a server discovery via the well-known document.

Two new fields are added to this document to support OIDC Provider discovery:

- REQUIRED `issuer` - the OIDC Provider that is trusted by the homeserver
- OPTIONAL `account` - the URL where the user is able to access the account management capabilities of the OIDC Provider

For example:

```
GET /.well-known/matrix/client
Host: example.com
Accept: application/json
```

```
HTTP/1.1 200 OK
Content-Type: application/json
```

```json
{
  "m.homeserver": {
    "base_url": "https://matrix-client.example.com"
  },
  "m.identity_server": {
    "base_url": "https://identity.example.com"
  },
  "m.authentication": {
    "issuer": "https://account.example.com",
    "account": "https://account.example.com/myaccount"
  }
}
```

> The authentication server metadata can then be discovered by the client using [OpenID Connect Discovery 1.0](https://openid.net/specs/openid-connect-discovery-1_0.html) against the `issuer` field.

Reviewer: How does "normal" (non-Matrix) OIDC handle getting this metadata? Is `/.well-known/openid-configuration` standard? Or is this just always left to the protocol to negotiate?

Reviewer: Looks like https://github.com/sandhose/matrix-doc/blob/msc/sandhose/oidc-discovery/proposals/2965-oidc-discovery.md#discovery-via-webfinger might answer my question.

```
GET /.well-known/openid-configuration
Host: account.example.com
Accept: application/json
```

```
HTTP/1.1 200 OK
Content-Type: application/json
```

```json
{
  "issuer": "https://account.example.com/",
  "authorization_endpoint": "https://account.example.com/oauth2/auth",
  "token_endpoint": "https://account.example.com/oauth2/token",
  "registration_endpoint": "https://account.example.com/oauth2/clients/register",

>   "end_session_edntpoint": "https://account.example.com/oauth2/logout",

Reviewer 2: Typo: `end_session_edntpoint` -> `end_session_endpoint`

  "jwks_uri": "https://account.example.com/.well-known/jwks.json",
  "response_types_supported": ["code"],
  "grant_types_sypported": ["authorization_code", "refresh_token"],
  "response_mode_sypported": ["query", "fragment"],
  "//": "some fields omitted"
}
```

The account management URL may accept the following additional query parameters:

> - RECOMMENDED `id_token_hint` - ID Token (as previously issued by the issuer to the client) as a hint about the current authenticated user that is requesting to be able to manage their account. This may be used by the issuer to: if not logged in then used as a login hint; if already logged in, but for a different user/identity then warn the user that they are accessing account.

Reviewer: I'm finding this hard to follow, maybe the following is clearer? (I'm not sure it is accurate.)

```suggestion
- RECOMMENDED `id_token_hint` - An ID Token that was previously issued to the client; the issuer uses it as a hint for which user is requesting to manage their account. If the requesting user is not logged in then it is used as a login hint; if a different user/identity is already logged in then warn the user that they are accessing a different account.
```

## Potential issues

Not all Matrix servers have the well-known client discovery mechanism setup.
Unlike the identity server, the authentication server is necessary to login against a Matrix server.

> Being unable to discover the authorization server from the Matrix Client API might be an issue in some cases.

Reviewer: This is a pretty big issue IMO. Can we fallback to a different location?

See also MSC3922?

Reviewer: And I see the Alternatives section below seems to cover this. 😄

### Discovery via a new client API endpoint

The Matrix Client API could have a new endpoint to discover the issuer, e.g. `/_matrix/client/r0/auth_issuer`.

### Discovery via the `m.login.oauth2` authentication method

The spec already defines a `m.login.oauth2` authentication method, but it was never implemented.
The downside of this approach is that the plan is to deprecate the old login mechanism and it does not make sense to keep it just to discover the issuer.

### Discovery via WebFinger

OIDC already has a standard way to discover OP from an identifier: WebFinger. This is already adopted by Mastodon, and might help solving logging in via 3PIDs like emails.

Sample exchange:

```
GET /.well-known/webfinger?
    resource= mxid:@john:example.com &
    rel= http://openid.net/specs/connect/1.0/issuer
Host: example.com
```

```json
{
  "subject": "mxid:@john:matrix.org",
  "links": [
    {
      "rel": "http://openid.net/specs/connect/1.0/issuer",
      "href": "https://account.example.com"
    }
  ]
}
```

The `mxid` scheme is a bit arbitrary here.
The parameters in the URL should be percent-encoded, this was left unencoded for clarity.

> The benefits of this approach are that it is standard and decouples the authentication server from the Matrix server: different authentication servers could be used by accounts on the server.

Reviewer: ```suggestion
The benefits of this approach are that it is standard and decouples the authentication server from the Matrix server: different authentication servers could be used by different accounts on the server.
```

The downsides of this approach are:

> - the `.well-known` resource is dynamic, which can be harder to host/delegate & might conflict with other services like Mastodon

Reviewer: An alternative viewpoint is that it might mean you only need to setup a single endpoint to handle authentication for multiple services.

It is dynamic because you need to return the `subject` back in the response?

Reviewer 3: (I think this is resolved with the rewrite - if not, please open a new thread)

> - this does not cover discovering the authentication server for user registration

Reviewer: Is this left to outside the OIDC standards?

Reviewer 3: (I think this is resolved with the rewrite - if not, please open a new thread)

# MSC2965: OIDC Provider discovery

This proposal is part of the broader [MSC3861: Matrix architecture change to delegate authentication via OIDC](https://github.com/matrix-org/matrix-spec-proposals/pull/3861).

To be able to initiate an OAuth 2.0 login flow to use a Matrix server, the client needs to know the location of the authentication server in use.

## Proposal

This introduces a new Client-Server API endpoint to discover the authentication server used by the homeserver.
It also introduces new OpenID Connect Provider metadata to allow clients to deep-link to the account management capabilities of the authentication server.

### `GET /auth_issuer`

A request on this endpoint should return a JSON object with one field:

- _REQUIRED_ `issuer`: the OpenID Connect Provider that is trusted by the homeserver

For example:

```http

> GET /_matrix/client/v3/auth_issuer

Reviewer 4: Why `v3`? It's a new endpoint, shouldn't it be `v1`?

# MSC2965: OIDC Provider discovery

This proposal is part of the broader [MSC3861: Matrix architecture change to delegate authentication via OIDC](https://github.com/matrix-org/matrix-spec-proposals/pull/3861).

To be able to initiate an OAuth 2.0 login flow to use a Matrix server, the client needs to know the location of the authentication server in use.

## Proposal

This introduces a new Client-Server API endpoint to discover the authentication server used by the homeserver.
It also introduces new OpenID Connect Provider metadata to allow clients to deep-link to the account management capabilities of the authentication server.

### `GET /auth_issuer`

A request on this endpoint should return a JSON object with one field:

> - _REQUIRED_ `issuer`: the OpenID Connect Provider that is trusted by the homeserver

Reviewer 5: what should homeservers that do not use OIDC do here? Return a 404? With a matrix error code, or not?

We should specify exactly what responses mean "I don't do OIDC", so that clients can distinguish servers that don't do OIDC from those that are unreachable or having a transient problem.

We also need to be mindful of backwards-compatibility, of course.

Reviewer 3: @sandhose this should probably be addressed, at least while the spec plans to support UIA (though deprecated) in the meantime.

Author: Have clarified that in 93d1b094

> - _REQUIRED_ `issuer`: the OpenID Connect Provider that is trusted by the homeserver

Reviewer 6: I think I haven't seen this addressed anywhere but why does the mapping need to be 1:1? Could a homeserver not trust and work with several OPs letting the user choose between them?

Reviewer 7: I agree. It is not uncommon to see multiple choices:
- Log in with GitHub
- Log in with Google
- Log in with ...

Reviewer 3: (I think this is resolved with the rewrite - if not, please open a new thread)
